gpii/node_modules/gpii-oauth2/gpii-oauth2-datastore/src/DbDataStoreUtils.js

```diff
@@ -0,0 +1,397 @@
+// of "_id".
+gpii.oauth2.dbDataStore.cleanUpDoc = function (data) {
+    if (data) {
+        data.id = data._id;
+        delete data._id;
+        delete data._rev;
+        delete data.type;
+    }
+    return data;
+};
+
+// Handle cases when multiple records to be returned in an array
+gpii.oauth2.dbDataStore.handleMultipleRecords = function (data) {
+    var records = [];
+    fluid.each(data, function (row) {
+        var oneRecord = gpii.oauth2.dbDataStore.cleanUpDoc(row.value);
+        records.push(oneRecord);
+    });
+    return records;
+};
+
+// Create a new record
+gpii.oauth2.dbDataStore.addRecord = function (dataSource, docType, idName, data) {
+    var promise = fluid.promise();
+    // TODO: Requires proper field validation on data instead of only checking against falsy data.
+    if (data) {
+        var directModel = {};
+        directModel[idName] = uuid.v4();
+        fluid.extend(data, {type: docType});
+        var finalDirectModel = $.extend(true, {}, dataSource.options.directModel, directModel);
+        promise = dataSource.set(finalDirectModel, data);
+    } else {
+        var error = gpii.oauth2.composeError(gpii.oauth2.errors.missingDoc, {docName: docType});
+        promise.reject(error);
+    }
+    return promise;
+};
+
+// Update an existing record
+gpii.oauth2.dbDataStore.updateRecord = function (dataSource, docType, idName, data) {
+    var directModel = {};
+    directModel[idName] = data.id;
+    fluid.extend(data, {type: docType});
+    delete data.id;
+    var finalDirectModel = $.extend(true, {}, dataSource.options.directModel, directModel);
+    var promise = dataSource.set(finalDirectModel, data);
+    return promise;
+};
+
+// Authorization Decisions
+// -----------------------
+
+// ==== updateAuthDecision()
+gpii.oauth2.dbDataStore.updateAuthDecision = function (that, userId, authDecisionData) {
+    var input = {
+        userId: userId,
+        authDecisionData: authDecisionData
+    };
+    return fluid.promise.fireTransformEvent(that.events.onUpdateAuthDecision, {inputArgs: input});
+};
+
+/*
+ * Handles the case where the resolved value of the inputPromise needs to trigger an error of missing documents.
+ * When the resolved value exists, process the value with the dataProcessFunc argument and resolved the processed
+ * resolve with promiseTogo argument.
+ * @inputPromise (Promise object): The input promise object to be processed.
+ * @input: The input data.
+ * @docType (String): The document type to be reported when the resolved value of inputPromise is undefined.
+ * @allowUndefinedValue: Allows the value resolved from the inputPromise to be undefined
+ *
+ * @return (Promise object): A promise object that carries the result of this function. The result is either a value
+ * that gets resolved into the returned promise or a rejection with a missing doc error.
+ */
+gpii.oauth2.dbDataStore.handlePromiseWithMissingDoc = function (inputPromise, input, docType, allowUndefinedValue) {
```

Comment:
`handlePromise` is a bad, generic name. Consult the masterwork "The Abstract is an Enemy" by Blackwell, Church and Green at http://www.ppig.org/papers/20th-blackwell.pdf . The user will have no idea what `inputPromise` is when they come to read this function - explain what workflow we are in, and what the function does for us, as well as updating the name